sunscreen/src/fhe_compiler.rs

```diff
@@ -46,117 +47,85 @@ pub trait FheProgramFn {
 /**
  * A frontend compiler for Sunscreen FHE programs.
  */
-pub struct Compiler {
+pub struct Compiler<T> {
     fhe_program_fns: Vec<Box<dyn FheProgramFn>>,
+    zkp_program_fns: Vec<Box<dyn ZkpProgramFn>>,
     params_mode: ParamsMode,
     plain_modulus_constraint: PlainModulusConstraint,
     security_level: SecurityLevel,
     noise_margin: u32,
+    _phantom: PhantomData<T>,
 }
 
-impl Default for Compiler {
+impl Default for Compiler<()> {
     fn default() -> Self {
         Self::new()
     }
 }
 
-impl Compiler {
+/**
+ * A private method for converting between Compiler types. We don't want
+ * to expose this as part of the public API, hence we don't `impl From`.
+ */
+fn into<T, U>(x: Compiler<T>) -> Compiler<U> {
```

Comment:
Does it make sense to have this exist for all `T, U`? It's internal so it's not a huge deal, but if only certain conversions make sense, we could restrict this at the type level.

Reply:
The cases this needs to handle are:
`Compiler<()> -> Compiler<Zkp>`
`Compiler<()> -> Compiler<Fhe>`
`Compiler<Zkp> -> Compiler<FheZkp>`
`Compiler<Fhe> -> Compiler<FheZkp>`

To add bounds to make all of these work, I think we would need to add a new trait and not use (), but some other marker type. I don't think it's worth it.

This function is indeed internal only.